### ISSUE TITLE
Immediate Dead Kernel -- timeout error

### ISSUE THREAD

**Issue by vnredmon** [Calysto/octave_kernel#12]
Hi, I am trying to get the octave kernel running within my ipython notebook and am getting an immediate dead kernel upon opening a new notebook. This is on a fresh virtualenv with ipython 3.2, octave_kernel 0.11.0, and metakernel 0.9.0.  
Do you have any thoughts on the cause? The terminal ipython output is below (long). 

Thanks.

```
(venv)➜  Documents  ipython notebook
[I 16:44:36.368 NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
[I 16:44:36.735 NotebookApp] The port 8888 is already in use, trying another random port.
[W 16:44:36.741 NotebookApp] Terminals not available (error was No module named terminado)
[I 16:44:36.741 NotebookApp] Serving notebooks from local directory: /Users/vnredmon/Documents
[I 16:44:36.741 NotebookApp] 0 active kernels
[I 16:44:36.742 NotebookApp] The IPython Notebook is running at: http://localhost:8889/
[I 16:44:36.742 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 16:44:42.104 NotebookApp] Creating new notebook in
[I 16:44:43.893 NotebookApp] Kernel started: 4857fa84-af7e-4b94-a5de-88b43ee165f6
[W 16:44:54.002 NotebookApp] Timeout waiting for kernel_info reply from 4857fa84-af7e-4b94-a5de-88b43ee165f6
[I 16:45:14.502 NotebookApp] Saving file at /Untitled.ipynb
[IPKernelApp] ERROR | Can't load '/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/magics/shell_magic.py': error: Timeout exceeded.
<metakernel.pexpect.spawnu object at 0x106a2b390>
version: 3.3
command: /bin/bash
args: ['/bin/bash']
searcher: None
buffer (last 100 chars): u'prompt_info: command not found\r\n\x1b[?1034h(venv)%{[green]%}%p %{[cyan]%}%c %{[blue]%}%{[blue]%} % %{%}'
before (last 100 chars): u'prompt_info: command not found\r\n\x1b[?1034h(venv)%{[green]%}%p %{[cyan]%}%c %{[blue]%}%{[blue]%} % %{%}'
after: <class 'metakernel.pexpect.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 64741
child_fd: 54
closed: False
timeout: 30
delimiter: <class 'metakernel.pexpect.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
---------------------------------------------------------------------------
EOF                                       Traceback (most recent call last)
/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.pyc in _run_module_as_main(mod_name='octave_kernel', alter_argv=1)
    160         sys.argv[0] = fname
    161     return _run_code(code, main_globals, None,
--> 162                      "__main__", fname, loader, pkg_name)
        fname = '/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc'
        loader = <zipimporter object "/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg">
        pkg_name = ''
    163
    164 def run_module(mod_name, init_globals=None,

/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.pyc in _run_code(code=<code object <module> at 0x1053ffb30, file "buil...acosx-10.10-x86_64/egg/octave_kernel.py", line 1>, run_globals={'IPKernelApp': <class 'IPython.kernel.zmq.kernelapp.IPKernelApp'>, 'Image': <class 'IPython.core.display.Image'>, 'MetaKernel': <class 'metakernel._metakernel.MetaKernel'>, 'OctaveKernel': <class '__main__.OctaveKernel'>, 'ProcessMetaKernel': <class 'metakernel.process_metakernel.ProcessMetaKernel'>, 'REPLWrapper': <class 'metakernel.replwrap.REPLWrapper'>, 'SVG': <class 'IPython.core.display.SVG'>, '__builtins__': <module '__builtin__' (built-in)>, '__doc__': None, '__file__': '/Users/vnredmon/Documents/venv/lib/python2.7/sit.../octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc', ...}, init_globals=None, mod_name='__main__', mod_fname='/Users/vnredmon/Documents/venv/lib/python2.7/sit.../octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc', mod_loader=<zipimporter object "/Users/vnredmon/Documents/v....7/site-packages/octave_kernel-0.11.0-py2.7.egg">, pkg_name='')
     70                        __loader__ = mod_loader,
     71                        __package__ = pkg_name)
---> 72     exec code in run_globals
        code = <code object <module> at 0x1053ffb30, file "build/bdist.macosx-10.10-x86_64/egg/octave_kernel.py", line 1>
        run_globals = {'OctaveKernel': <class '__main__.OctaveKernel'>, 'SVG': <class 'IPython.core.display.SVG'>, '__version__': '0.11.0', 'ProcessMetaKernel': <class 'metakernel.process_metakernel.ProcessMetaKernel'>, '__builtins__': <module '__builtin__' (built-in)>, 'Image': <class 'IPython.core.display.Image'>, '__file__': '/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc', 'subprocess': <module 'subprocess' from '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.pyc'>, '__package__': '', 'sys': <module 'sys' (built-in)>, 'tempfile': <module 'tempfile' from '/usr/local/Cellar/python/2.7.10/Frameworks/Python.framework/Versions/2.7/lib/python2.7/tempfile.pyc'>, 'MetaKernel': <class 'metakernel._metakernel.MetaKernel'>, 'u': <function u at 0x106249c08>, 'IPKernelApp': <class 'IPython.kernel.zmq.kernelapp.IPKernelApp'>, 'REPLWrapper': <class 'metakernel.replwrap.REPLWrapper'>, '__name__': '__main__', '__loader__': <zipimporter object "/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg">, 'os': <module 'os' from '/Users/vnredmon/Documents/venv/lib/python2.7/os.pyc'>, '__doc__': None, 'print_function': _Feature((2, 6, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0), 65536)}
     73     return run_globals
     74

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc in <module>()

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/config/application.pyc in launch_instance(cls=<class 'IPython.kernel.zmq.kernelapp.IPKernelApp'>, argv=None, **kwargs={'kernel_class': <class '__main__.OctaveKernel'>})
    571         """
    572         app = cls.instance(**kwargs)
--> 573         app.initialize(argv)
        app.initialize = <bound method IPKernelApp.initialize of <IPython.kernel.zmq.kernelapp.IPKernelApp object at 0x1062a4590>>
        argv = None
    574         app.start()
    575

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/kernel/zmq/kernelapp.pyc in initialize(self=<IPython.kernel.zmq.kernelapp.IPKernelApp object>, argv=None)

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/config/application.pyc in catch_config_error(method=<function initialize>, app=<IPython.kernel.zmq.kernelapp.IPKernelApp object>, *args=(None,), **kwargs={})
     73     """
     74     try:
---> 75         return method(app, *args, **kwargs)
        method = <function initialize at 0x1062a8758>
        app = <IPython.kernel.zmq.kernelapp.IPKernelApp object at 0x1062a4590>
        args = (None,)
        kwargs = {}
     76     except (TraitError, ArgumentError) as e:
     77         app.print_help()

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/kernel/zmq/kernelapp.pyc in initialize(self=<IPython.kernel.zmq.kernelapp.IPKernelApp object>, argv=None)
    353         self.init_io()
    354         self.init_signal()
--> 355         self.init_kernel()
        self.init_kernel = <bound method IPKernelApp.init_kernel of <IPython.kernel.zmq.kernelapp.IPKernelApp object at 0x1062a4590>>
    356         # shell init steps
    357         self.init_path()

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/kernel/zmq/kernelapp.pyc in init_kernel(self=<IPython.kernel.zmq.kernelapp.IPKernelApp object>)
    308                                 log=self.log,
    309                                 profile_dir=self.profile_dir,
--> 310                                 user_ns=self.user_ns,
        global user_ns = undefined
        self.user_ns = None
    311         )
    312         kernel.record_ports(self.ports)

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/IPython/config/configurable.pyc in instance(cls=<class '__main__.OctaveKernel'>, *args=(), **kwargs={'iopub_socket': <zmq.sugar.socket.Socket object>, 'log': <logging.Logger object>, 'parent': <IPython.kernel.zmq.kernelapp.IPKernelApp object>, 'profile_dir': <IPython.core.profiledir.ProfileDir object>, 'session': <IPython.kernel.zmq.session.Session object>, 'shell_streams': [<zmq.eventloop.zmqstream.ZMQStream object>, <zmq.eventloop.zmqstream.ZMQStream object>], 'stdin_socket': <zmq.sugar.socket.Socket object>, 'user_ns': None})
    335         # Create and save the instance
    336         if cls._instance is None:
--> 337             inst = cls(*args, **kwargs)
        inst = undefined
        cls = <class '__main__.OctaveKernel'>
        args = ()
        kwargs = {'log': <logging.Logger object at 0x1062a4ad0>, 'parent': <IPython.kernel.zmq.kernelapp.IPKernelApp object at 0x1062a4590>, 'shell_streams': [<zmq.eventloop.zmqstream.ZMQStream object at 0x1062c75d0>, <zmq.eventloop.zmqstream.ZMQStream object at 0x1062c7850>], 'user_ns': None, 'iopub_socket': <zmq.sugar.socket.Socket object at 0x10628f9a8>, 'profile_dir': <IPython.core.profiledir.ProfileDir object at 0x1062c7310>, 'session': <IPython.kernel.zmq.session.Session object at 0x1062a4f50>, 'stdin_socket': <zmq.sugar.socket.Socket object at 0x10628f940>}
    338             # Now make sure that the instance will also be returned by
    339             # parent classes' _instance attribute.

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/process_metakernel.pyc in __init__(self=<__main__.OctaveKernel object>, **kwargs={'iopub_socket': <zmq.sugar.socket.Socket object>, 'log': <logging.Logger object>, 'parent': <IPython.kernel.zmq.kernelapp.IPKernelApp object>, 'profile_dir': <IPython.core.profiledir.ProfileDir object>, 'session': <IPython.kernel.zmq.session.Session object>, 'shell_streams': [<zmq.eventloop.zmqstream.ZMQStream object>, <zmq.eventloop.zmqstream.ZMQStream object>], 'stdin_socket': <zmq.sugar.socket.Socket object>, 'user_ns': None})
     40         self.wrapper = None
     41         self.repr = str
---> 42         self._start()
        self._start = <bound method OctaveKernel._start of <__main__.OctaveKernel object at 0x1062c7990>>
     43
     44     def _start(self):

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/process_metakernel.pyc in _start(self=<__main__.OctaveKernel object>)
     45         if not self.wrapper is None:
     46             self.wrapper.child.terminate()
---> 47         self.wrapper = self.makeWrapper()
        self.wrapper = None
        self.makeWrapper = <bound method OctaveKernel.makeWrapper of <__main__.OctaveKernel object at 0x1062c7990>>
     48
     49     def do_execute_direct(self, code):

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/octave_kernel-0.11.0-py2.7.egg/octave_kernel.pyc in makeWrapper(self=<__main__.OctaveKernel object>)

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/replwrap.pyc in __init__(self=<metakernel.replwrap.REPLWrapper object>, cmd_or_spawn='octave', prompt_regex=u'octave.*>', prompt_change_cmd=u"PS1('{0}'); PS2('{1}')", new_prompt_regex=u'PEXPECT_PROMPT>', continuation_prompt_regex=u'PEXPECT_PROMPT_', extra_init_cmd=None, prompt_emit_cmd=None, echo=False)
     73             self.set_prompt(prompt_regex,
     74                             prompt_change_cmd.format(new_prompt_regex,
---> 75                                                      continuation_prompt_regex))
        continuation_prompt_regex = u'PEXPECT_PROMPT_'
     76             self.prompt_regex = new_prompt_regex
     77         self.continuation_prompt_regex = continuation_prompt_regex

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/replwrap.pyc in set_prompt(self=<metakernel.replwrap.REPLWrapper object>, prompt_regex=u'octave.*>', prompt_change_cmd=u"PS1('PEXPECT_PROMPT>'); PS2('PEXPECT_PROMPT_')")
     90
     91     def set_prompt(self, prompt_regex, prompt_change_cmd):
---> 92         self.child.expect(prompt_regex)
        self.child.expect = <bound method spawnu.expect of <metakernel.pexpect.spawnu object at 0x1062fe390>>
        prompt_regex = u'octave.*>'
     93         self.sendline(prompt_change_cmd)
     94

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/pexpect.pyc in expect(self=<metakernel.pexpect.spawnu object>, pattern=u'octave.*>', timeout=-1, searchwindowsize=-1, async=False)
   1647         compiled_pattern_list = self.compile_pattern_list(pattern)
   1648         return self.expect_list(compiled_pattern_list,
-> 1649                 timeout, searchwindowsize, async)
        timeout = -1
        searchwindowsize = -1
        async = False
   1650
   1651     def expect_list(self, pattern_list, timeout=-1, searchwindowsize=-1,

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/pexpect.pyc in expect_list(self=<metakernel.pexpect.spawnu object>, pattern_list=[<_sre.SRE_Pattern object>], timeout=30, searchwindowsize=-1, async=False)
   1668
   1669         exp = Expecter(self, searcher_re(pattern_list), searchwindowsize)
-> 1670         return exp.expect_loop(timeout)
        exp.expect_loop = <bound method Expecter.expect_loop of <metakernel.pexpect.Expecter object at 0x106a2b410>>
        timeout = 30
   1671
   1672     def expect_exact(self, pattern_list, timeout=-1, searchwindowsize=-1,

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/pexpect.pyc in expect_loop(self=<metakernel.pexpect.Expecter object>, timeout=29.981853008270264)
    228                     timeout = end_time - time.time()
    229         except EOF as e:
--> 230             return self.eof(e)
        self.eof = <bound method Expecter.eof of <metakernel.pexpect.Expecter object at 0x106a2b410>>
        e = EOF('End Of File (EOF). Empty string style platform.',)
    231         except TIMEOUT as e:
    232             return self.timeout(e)

/Users/vnredmon/Documents/venv/lib/python2.7/site-packages/metakernel-0.9.0-py2.7.egg/metakernel/pexpect.pyc in eof(self=<metakernel.pexpect.Expecter object>, err=EOF('End Of File (EOF). Empty string style platform.',))
    177             if err is not None:
    178                 msg = str(err) + '\n' + msg
--> 179             raise EOF(msg)
        global EOF = <class 'metakernel.pexpect.EOF'>
        msg = "End Of File (EOF). Empty string style platform.\n<metakernel.pexpect.spawnu object at 0x1062fe390>\nversion: 3.3\ncommand: /usr/local/bin/octave\nargs: ['/usr/local/bin/octave']\nsearcher: None\nbuffer (last 100 chars): u''\nbefore (last 100 chars): u'\\n  Referenced from: /usr/local/Cellar/octave/3.8.2/bin/octave-cli-3.8.2\\r\\n  Reason: image not found\\r\\n'\nafter: <class 'metakernel.pexpect.EOF'>\nmatch: None\nmatch_index: None\nexitstatus: None\nflag_eof: True\npid: 64764\nchild_fd: 54\nclosed: False\ntimeout: 30\ndelimiter: <class 'metakernel.pexpect.EOF'>\nlogfile: None\nlogfile_read: None\nlogfile_send: None\nmaxread: 2000\nignorecase: False\nsearchwindowsize: None\ndelaybeforesend: 0.05\ndelayafterclose: 0.1\ndelayafterterminate: 0.1"
    180
    181     def timeout(self, err=None):

EOF: End Of File (EOF). Empty string style platform.
<metakernel.pexpect.spawnu object at 0x1062fe390>
version: 3.3
command: /usr/local/bin/octave
args: ['/usr/local/bin/octave']
searcher: None
buffer (last 100 chars): u''
before (last 100 chars): u'\n  Referenced from: /usr/local/Cellar/octave/3.8.2/bin/octave-cli-3.8.2\r\n  Reason: image not found\r\n'
after: <class 'metakernel.pexpect.EOF'>
match: None
match_index: None
exitstatus: None
flag_eof: True
pid: 64764
child_fd: 54
closed: False
timeout: 30
delimiter: <class 'metakernel.pexpect.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
```

**Comment by vnredmon** [comment]
Ok, I'm an idiot. Something screwed up my octave installation. 

Closing this.

**Comment by blink1073** [comment]
Ha, glad to be of service :smile:.

**Comment by BielStela** [comment]
I have the same problem. What did you did for fixing it?

**Comment by blink1073** [comment]
Hi @BielStela, what is your exact output?

**Comment by BielStela** [comment]
hi, this is my output: 

```
[I 16:42:54.711 NotebookApp] KernelRestarter: restarting kernel (4/5)
WARNING:root:kernel 8ba15733-58ad-4957-a4db-2fc895731ba3 restarted
[IPKernelApp] ERROR | Can't load 'C:\Users\Biel\Anaconda2\lib\site-packages\metakernel\magics\shell_magic.py': error: End Of File (EOF).
<pexpect.popen_spawn.PopenSpawn object at 0x0000000004710EB8>
searcher: searcher_re:
    0: re.compile("__repl_ready__")
    1: re.compile("PEXPECT_PROMPT_")
    2: re.compile("PEXPECT_PROMPT+")
Traceback (most recent call last):
  File "C:\Users\Biel\Anaconda2\lib\runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "C:\Users\Biel\Anaconda2\lib\runpy.py", line 72, in _run_code
    exec code in run_globals
  File "C:\Users\Biel\Anaconda2\lib\site-packages\octave_kernel\__main__.py", line 3, in <module>
    IPKernelApp.launch_instance(kernel_class=OctaveKernel)
  File "C:\Users\Biel\Anaconda2\lib\site-packages\traitlets\config\application.py", line 652, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-121>", line 2, in initialize
  File "C:\Users\Biel\Anaconda2\lib\site-packages\traitlets\config\application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "C:\Users\Biel\Anaconda2\lib\site-packages\ipykernel\kernelapp.py", line 454, in initialize
    self.init_kernel()
  File "C:\Users\Biel\Anaconda2\lib\site-packages\ipykernel\kernelapp.py", line 365, in init_kernel
    user_ns=self.user_ns,
  File "C:\Users\Biel\Anaconda2\lib\site-packages\traitlets\config\configurable.py", line 412, in instance
    inst = cls(*args, **kwargs)
  File "C:\Users\Biel\Anaconda2\lib\site-packages\metakernel\process_metakernel.py", line 52, in __init__
    self._start()
  File "C:\Users\Biel\Anaconda2\lib\site-packages\metakernel\process_metakernel.py", line 57, in _start
    self.wrapper = self.makeWrapper()
  File "C:\Users\Biel\Anaconda2\lib\site-packages\octave_kernel\kernel.py", line 56, in makeWrapper
    return self.octave_engine.repl
  File "C:\Users\Biel\Anaconda2\lib\site-packages\octave_kernel\kernel.py", line 50, in octave_engine
    stream_handler=self.Print)
  File "C:\Users\Biel\Anaconda2\lib\site-packages\octave_kernel\kernel.py", line 97, in __init__
    self.repl = self._create_repl()
  File "C:\Users\Biel\Anaconda2\lib\site-packages\octave_kernel\kernel.py", line 277, in _create_repl
    version = subprocess.check_output(version_cmd).decode('utf-8')
  File "C:\Users\Biel\Anaconda2\lib\subprocess.py", line 567, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "C:\Users\Biel\Anaconda2\lib\subprocess.py", line 711, in __init__
    errread, errwrite)
  File "C:\Users\Biel\Anaconda2\lib\subprocess.py", line 959, in _execute_child
    startupinfo)
WindowsError: [Error 5] Access is denied
[W 16:42:57.721 NotebookApp] KernelRestarter: restart failed
[W 16:42:57.723 NotebookApp] Kernel 8ba15733-58ad-4957-a4db-2fc895731ba3 died, removing from map.
ERROR:root:kernel 8ba15733-58ad-4957-a4db-2fc895731ba3 restarted failed!
[W 16:42:57.779 NotebookApp] Kernel deleted before session 
[W 16:42:57.780 NotebookApp] 410 DELETE /api/sessions/7d00be28-476b-4892-9e26-fab603a5e8d6 (::1) 3.00ms referer=http://localhost:8888/notebooks/Documents/NN%20course/Untitled.ipynb?kernel_name=octave
```

**Comment by blink1073** [comment]
Drat, it looks like it is getting stuck trying to start the `%shell` magic at startup.  I hadn't noticed because I have `bash` on `Windows`.  This will take a patch to `metakernel`.  I should be able to fix this over the weekend.

**Comment by blink1073** [comment]
Hi @BielStela, sorry, I lost track of this one because the issue is closed.  Can you please try with the latest `octave_kernel` (`pip install -U octave_kernel`)?  I just tried on Windows 10 with no bash, and it loads cmd just fine.  I think changes we made recently to metakernel have taken care of it.  If you're still seeing the problem, please open a new issue.
